### PR TITLE
[improving] store grpc addr to node info and auto refresh backends addrs for grpc client

### LINKED ISSUE #5308: [metasrv] store grpc_addr automatically when a new metasrv node launching.
**Summary**
Store grpc_addr automatically when a new metasrv node launching and remove it when node shutdown, so the grpc client could fetch all available meta-grpc backend addrs for connection.

### PR DESCRIPTION
**Pull request by ariesdevil**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

* [x] Store grpc addr to node info
* [x] auto-refresh backends addrs for grpc client
* [x] tests

## Changelog

- New Feature
- Improvement

## Related Issues

Fixes #5308

### CODE REVIEW

File: common/meta/types/src/meta_errors.rs [hunk 1]
```diff
@@ -82,7 +82,7 @@ pub enum MetaError {
     Fatal(AnyError),
 }
 
-pub type MetaResult<T> = std::result::Result<T, MetaError>;
+pub type MetaResult<T> = Result<T, MetaError>;
```
Comment (Xuanwo) [inline common/meta/types/src/meta_errors.rs:85]: :+1:

File: metasrv/src/bin/metasrv.rs [hunk 2]
```diff
@@ -75,14 +75,16 @@ async fn main(_global_tracker: Arc<RuntimeTracker>) -> common_exception::Result<
...
+            conf.grpc_api_address.clone()
         );
         srv.start().await.expect("Databend meta service error");
         stop_handler.push(Box::new(srv));
     }
 
     // join raft cluster after all service started
-    meta_node.join_cluster(&conf.raft_config).await?;
+    meta_node
+        .join_cluster(&conf.raft_config, conf.grpc_api_address)
```
Comment (Xuanwo) [inline metasrv/src/bin/metasrv.rs:86]: I'm confused about the `raft_config`.

Will we use `raft_config` outside `metasrv`? How about merging them into `MetaConfig` directly?

File: query/src/common/meta/meta_client.rs [hunk 3]
```diff
@@ -46,7 +46,7 @@ impl MetaClientProvider {
 
     /// Get kv async client, operations trait defined in KVApi.
     pub async fn try_get_kv_client(&self) -> Result<Arc<dyn KVApi>> {
-        let local = self.grpc_conf.meta_service_config.address.is_empty();
+        let local = self.grpc_conf.kv_service_config.address.is_empty();
```
Comment (Xuanwo) [inline query/src/common/meta/meta_client.rs:49]: Considering https://github.com/datafuselabs/databend/issues/5323

Will we adopt `kv_service_config` always and remove `meta_service_config`?

**Comment by Xuanwo** [comment]
`test_metactl` failed:

```shell
Error: missing field `grpc_api_addr` at line 1 column 195
```

Seems we are introducing changes that are not compatible.

File: metasrv/src/bin/metasrv.rs (hunk 2 above)
Comment (ariesdevil) [reply metasrv/src/bin/metasrv.rs:86]: `raft_config` is all about raft related config, `MetaConfig` includes other configs that nothing to do with raft. So I think I think it's better to separate.

Comment (drmingdrmer) [reply metasrv/src/bin/metasrv.rs:88]: Right...It's a bit dirty:

RaftConfig is also used by `common/meta/embedded`, which is an embedded meta server for testing purposes.

**Comment by drmingdrmer** [comment]
This looks like a breaking change: An old metasrv will crash loading the data written by a new metasrv that has `grpc_api_addr` in a `Node` record.

What about making it an `Option`?
```rust
pub struct Node {
    pub name: String,
    pub endpoint: Endpoint,
    pub grpc_api_addr: Option<String>, // <--- 🤔
}
```

**Comment by drmingdrmer** [comment]
Everything else looks good to me :D

**Comment by drmingdrmer** [comment]
If this PR is still under development, convert it to draft?

File: common/meta/grpc/src/grpc_client.rs [hunk 4]
```diff
@@ -346,7 +366,18 @@ impl MetaGrpcClient {
...
-                    resp.map(message::Response::MakeClient)
+                    match resp {
+                        Ok(r) => Ok(message::Response::MakeClient(r)),
+                        Err(e) => {
+                            tracing::warn!("make client error, try to remake. {:?}", e);
+                            let res = self.make_client().await;
+                            res.map(message::Response::MakeClient)
+                        }
+                    }
+                }
```
Comment (drmingdrmer) [inline common/meta/grpc/src/grpc_client.rs:377]: Isn't it better to let `self.make_client()` do the retry-loop when failed to build a client?

File: common/meta/grpc/src/grpc_client.rs [hunk 5]
```diff
@@ -491,25 +474,124 @@ impl MetaGrpcClient {
...
+            *current_endpoint = None;
+        }
+
+        let mut eps = self.get_endpoints().await;
+        debug_assert!(!eps.is_empty());
+
+        // remove unhealthy endpoints
+        {
+            let ues = self.unhealthy_endpoints.lock().unwrap();
+            eps.retain(|e| !ues.contains_key(e));
```
Comment (drmingdrmer) [inline common/meta/grpc/src/grpc_client.rs:489]: Is `retain()` a method of `TtlHashMap`?

File: common/meta/grpc/src/grpc_client.rs [hunk 6]
```diff
@@ -491,25 +474,124 @@ impl MetaGrpcClient {
...
     pub async fn set_endpoints(
         &self,
         endpoints: Vec<String>,
     ) -> std::result::Result<(), MetaError> {
         Self::endpoints_non_empty(&endpoints)?;
 
+        let distinct_cnt = endpoints.iter().filter(|n| !(*n).is_empty()).count();
+        if distinct_cnt < endpoints.len() / 2 + 1 {
+            tracing::warn!(
+                "distinct endpoints small than majority of meta cluster nodes {}<{}, endpoints: {:?}",
```
Comment (drmingdrmer) [inline common/meta/grpc/src/grpc_client.rs:550]: It may need a comment explaining why it does so

File: common/meta/grpc/src/grpc_client.rs [hunk 7]
```diff
@@ -491,25 +474,124 @@ impl MetaGrpcClient {
...
+        Ok(())
+    }
+
+    async fn auto_sync_endpoints(self: Arc<Self>, mut cancel_tx: OneSend<()>) {
+        loop {
+            select! {
+                _ = cancel_tx.closed() => {
+                    return;
+                }
+                _ = sleep(Duration::from_secs(5)) => {
```
Comment (drmingdrmer) [inline common/meta/grpc/src/grpc_client.rs:584]: `endpoints` changes rarely, hence the sleep time could be relatively longer than several seconds IMHO.

File: common/meta/grpc/src/grpc_client.rs [hunk 8]
```diff
@@ -775,10 +848,24 @@ impl MetaGrpcClient {
...
+        {
+            let ca = self.current_endpoint.lock().unwrap();
+            let mut ue = self.unhealthy_endpoints.lock().unwrap();
+            ue.insert((*ca).as_ref().unwrap().clone(), ());
+        }
+        {
+            let mut token = self.token.write().await;
+            *token = None;
+        }
+    }
```
Comment (drmingdrmer) [inline common/meta/grpc/src/grpc_client.rs:861]: Naming a method with what it does would be better than with what it does for.

E.g., to me, (maybe unthoughtful name) `mark_as_unhealthy()` would be better than `preapre_retry_req()`

File: common/meta/grpc/src/grpc_client.rs (hunk 5 above)
Comment (ariesdevil) [reply common/meta/grpc/src/grpc_client.rs:493]: nop, it's a `Vec`'s method

File: common/meta/grpc/src/grpc_client.rs (hunk 4 above)
Comment (ariesdevil) [reply common/meta/grpc/src/grpc_client.rs:381]: Good idea

File: common/meta/grpc/src/grpc_client.rs [hunk 9]
```diff
@@ -158,6 +164,10 @@ pub struct MetaGrpcClient {
 pub struct ClientHandle {
     /// For sending request to meta-client worker.
     pub(crate) req_tx: Sender<message::ClientWorkerRequest>,
+    /// Notify auto sync to stop.
+    /// `oneshot::Receiver` impl `Drop` by sending a closed notification to the `Sender` half.
+    #[allow(dead_code)]
+    cancel_auto_sync_rx: OneRecv<()>,
```
[the remark this answers is not in the document]
Comment (ariesdevil) [reply common/meta/grpc/src/grpc_client.rs:171]: it's still a dead_code since no explicitly used

File: common/meta/types/src/cluster.rs [hunk 10]
```diff
@@ -38,11 +38,21 @@ pub struct Slot {
...
 }
 
 impl fmt::Display for Node {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "{}={}", self.name, self.endpoint)
+        let grpc_addr_display = if self.grpc_api_addr.is_some() {
+            self.grpc_api_addr.clone().unwrap()
+        } else {
+            "".to_string()
+        };
```
Comment (drmingdrmer) [inline common/meta/types/src/cluster.rs:50]: `let x = if let Some(v) = ... { A } else { B }`

File: common/meta/grpc/src/grpc_client.rs [hunk 11]
```diff
@@ -775,10 +878,24 @@ impl MetaGrpcClient {
...
+    async fn mark_as_unhealthy(&self) {
+        {
+            let ca = self.current_endpoint.lock();
+            let mut ue = self.unhealthy_endpoints.lock();
+            ue.insert((*ca).as_ref().unwrap().clone(), ());
+        }
+        {
+            let mut token = self.token.write().await;
+            *token = None;
+        }
```
Comment (drmingdrmer) [inline common/meta/grpc/src/grpc_client.rs:890]: There is a concurrency issue that before 888, another task created a client, then the current task reset the token to None, which belongs to another connection.

File: common/meta/grpc/src/grpc_client.rs [hunk 12]
```diff
@@ -498,18 +559,69 @@ impl MetaGrpcClient {
...
+            .await?;
+        let result: Vec<String> = endpoints.into_inner().data;
+        self.set_endpoints(result).await?;
+        Ok(())
+    }
+
+    async fn auto_sync_endpoints(self: Arc<Self>, mut cancel_tx: OneSend<()>) {
+        if self.auto_sync_interval == Duration::from_secs(0) {
+            return;
+        }
```
Comment (drmingdrmer) [inline common/meta/grpc/src/grpc_client.rs:609]:
```
if let Some(interval) = self.auto_sync_interval {
}
```

File: common/meta/grpc/src/grpc_client.rs [hunk 13]
```diff
@@ -418,77 +446,110 @@ impl MetaGrpcClient {
...
+    }
 
-            item?
+    async fn make_channel(&self, addr: Option<&String>) -> std::result::Result<Channel, MetaError> {
+        let addr = if addr.is_some() {
+            addr.unwrap().clone()
+        } else {
+            let eps = self.endpoints.read().await;
+            eps.first().unwrap().clone()
         };
```
Comment (drmingdrmer) [inline common/meta/grpc/src/grpc_client.rs:529]: Use `if let Some(...)`

**Review by drmingdrmer — APPROVED** [review]

File: metasrv/src/meta_service/raftmeta.rs [hunk 14]
```diff
@@ -503,17 +504,17 @@ impl MetaNode {
                 body: ForwardRequestBody::Join(JoinRequest {
                     node_id: conf.id,
                     endpoint: raft_api_advertise_host_endpoint.clone(),
+                    grpc_api_addr: grpc_api_addr.clone(),
```
Comment (lichuang) [inline metasrv/src/meta_service/raftmeta.rs:507]: grpc_api_addr seems no more used, so can remove .clone()?

Comment (ariesdevil) [reply metasrv/src/meta_service/raftmeta.rs:507]: This is an assignment in a for loop, so need to clone here.

File: metasrv/src/meta_service/raftmeta.rs [hunk 15]
```diff
@@ -691,7 +702,18 @@ impl MetaNode {
...
         let sm = self.sto.state_machine.read().await;
-        sm.get_metasrv_addrs()
+        let nodes = sm.get_nodes()?;
+        let endpoints: Vec<String> = nodes
+            .iter()
+            .map(|n| {
+                if let Some(addr) = n.grpc_api_addr.clone() {
+                    addr
+                } else {
+                    "".to_string()
```
[the remark this answers is not in the document]
Comment (ariesdevil) [reply metasrv/src/meta_service/raftmeta.rs:712]: Here is for compatibility with older versions, the client will handle null values. I'll add some comments here.

**Comment by BohuTANG** [comment]
We need a benchmark test for this patch, and a check will it works well or not. cc @ariesdevil @drmingdrmer

**Review by lichuang — APPROVED** [review]

**Pull request merged**